ML-KEM: OpenSSL Part 1

vcsjones opened the pull request:
This brings in the minimum implementation to have a testable implementation of ML-KEM on Linux.

What is in this PR:

* Key generation
* Key import (seed, encapsulation key, decapsulation key)
* Key export (seed, encapsulation key, decapsulation key)
* Encapsulate and Decapsulate
* Microsoft.Bcl.Cryptography (only working implementation for now is type-forwards, down level always has `IsSupported` as `false` until a Windows implementation is available)
* Tests

What is not in this PR:

* PKCS8 / SPKI loading and export
* Convenience APIs, Callee-allocated buffers
* Platform-specific derived types (`MLKemOpenSsl`, etc)
* Certificates

Contributes to #113508

vcsjones committed:
Initial scaffolding and algorithm fetching

vcsjones committed:
Key generation

vcsjones committed:
Allow importing seeds into ML-KEM

vcsjones committed:
Encapsulate

vcsjones committed:
Public fields. Maybe a good idea

vcsjones committed:
Export decapsulation key

vcsjones committed:
Export encaps key

vcsjones committed:
NIST vectors for generation

vcsjones committed:
Minor tweaks

vcsjones committed:
Round trip on encapsulation and decapsulation

vcsjones committed:
Reorder

vcsjones committed:
Merge remote-tracking branch 'ms/main' into ml-kem

vcsjones committed:
Rework some key import and export things

vcsjones committed:
Encapsulation roundtrip

vcsjones committed:
Import Decapsulation key

vcsjones committed:
Check encapsulation key

vcsjones committed:
More tests and resource strings

vcsjones committed:
Keep working on tests

vcsjones committed:
Actual encapsulation tests

vcsjones committed:
Get tests building in OOB

vcsjones committed:
Improve test execution time

vcsjones committed:
Remove redundant MLKem from member names

vcsjones committed:
Add tests for MLKemAlgorithm

vcsjones committed:
Test and prohibit overlapping buffers

vcsjones committed:
Fixup docs

vcsjones committed:
Fix doc contents

vcsjones committed:
Remove duplicate define

vcsjones committed:
Fix error queue handling

vcsjones committed:
Fix build downlevel

vcsjones committed:
Merge branch 'main' into ml-kem

vcsjones committed:
Fixup compat

PranavSenthilnathan commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MLKemImplementation.OpenSsl.cs:
```diff
@@ -0,0 +1,123 @@
+        internal static MLKem GenerateKeyImpl(MLKemAlgorithm algorithm)
+        {
+            Debug.Assert(IsSupported);
+            SafeEvpKemHandle handle = MapAlgorithmToHandle(algorithm); // Shared handle, do not dispose.
```
Is it necessary to have this handle? It looks like it's mainly just used in key generation/import native code to get the name back, so couldn't we do the mapping on the C# side and pass in the name directly instead?

vcsjones replied:
I can get rid of the handle in the case of ML-KEM. The important part is correctly identifying if the algorithm is supported or not via fetch, but we don't need to keep the handle for that. We can just introduce a map.

vcsjones committed:
Remove held EVP-KEM handles.

vcsjones committed:
Make key importing generic for any algorithm using fromdata

vcsjones commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,411 @@
+
+        EVP_PKEY_CTX* ctx = NULL;
+        EVP_PKEY* pkey = NULL;
+        ctx = EVP_PKEY_CTX_new_from_name(NULL, name, NULL);
```
For importing, sure. For generation I would probably like to keep that ML-KEM specific, since it deals with an ML-KEM specific parameter.

vcsjones committed:
Remove experimental downlevel

vcsjones committed:
Simplify ML-KEM availability check.

The error queue from OpenSSL does not appear to be stable between OpenSSL versions. For now we can assume that if the fetch fails, its because the algorithm is not available.

vcsjones committed:
Re-add ExperimentalAttribute as internal on downlevel platforms

krwq commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:
```diff
@@ -22,7 +22,12 @@ namespace System.Diagnostics.CodeAnalysis
                     AttributeTargets.Interface |
                     AttributeTargets.Delegate, Inherited = false)]
-    public sealed class ExperimentalAttribute : Attribute
+    #if SYSTEM_PRIVATE_CORELIB
```
indent

krwq commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,301 @@
+
+        if (sharedSecretLengthT != Int32ToSizeT(sharedSecretLength))
+        {
+            ret = -1;
```
isn't this no-op given line below you set it back to 1

vcsjones replied:
That was missing a `goto done`. Thanks for the catch!

krwq commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,301 @@
+
+#include <assert.h>
+
+static int32_t GetKeyOctetStringParam(const EVP_PKEY* pKey,
```
do we need this to be a separate subroutine? If there is a benefit we could share this code with https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Security.Cryptography.Native/pal_ecc_import_export.c#L867 and all similar calls

vcsjones replied:
I'm not exactly sure what you are proposing, other than noting there are two things that use `EVP_PKEY_get_octet_string_param`.

In general I am a little wary of trying to do so much deduplication that it reaches the point of inscrutability. If there is something sensible we can do perhaps it can be done in a follow up PR.

krwq commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,301 @@
+        ERR_clear_error();
+
+        EVP_PKEY_CTX* ctx = NULL;
+        ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pKey, NULL);
```
Actually thinking about this, I don't know why would someone not want to use default implementation🤦

krwq commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Keys.cs:
```diff
@@ -0,0 +1,698 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Looking at dir structure I think this should be under `AlgorithmImplementations` dir

vcsjones replied:
Yeah it could be, though to be honest I don't see why we have the structure that way. I have been trying to simplify things over the years (trying to get rid of `Internal.Cryptography` and trying to reduce the amount of deep directory nesting.

If we feel strongly about continuing to follow this directory structure I can move it.

bartonjs replied:
I think that's a legacy from pre-1.0, when RSA and DSA weren't even in the same assembly.  (They got merged into Algorithms during 1.0).

Some directories are good for reducing clutter.. but since the src project is just grouped by namespace, and all the stuff lives in S.S.C, I'm supportive of "let's try something new" with these.

krwq commented on src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EvpPkey.cs:
```diff
@@ -19,6 +19,27 @@ internal static partial class Crypto
+            int keyLength,
+            [MarshalAs(UnmanagedType.Bool)] bool privateKey);
+
+        internal static SafeEvpPKeyHandle EvpPKeyFromData(string kemName, ReadOnlySpan<byte> key, bool privateKey)
```
since you're making this generic, you should probably rename `kemName`

vcsjones committed:
Code review feedback.

krwq commented on src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EVP.KemAlgs.cs:
```diff
@@ -0,0 +1,55 @@
+                // Do not use property initializers for these because we need to ensure CryptoInitializer.Initialize
+                // is called first. Property initializers happen before cctors, so instead set the property after the
+                // initializer is run.
+                MlKem512 = EvpKemAvailable(MLKemAlgorithm.MLKem512.Name);
```
I see we already do this elsewhere so I fine I guess

vcsjones replied:
The important bit is making sure we do the crypto initializer. Failing to do that results in hard to troubleshoot bugs like #100419. So I am inclined to keep following the established pattern.

vcsjones committed:
Fix typos.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLKem.cs:
```diff
@@ -0,0 +1,453 @@
+        /// <summary>
+        /// The shared secret size for ML-KEM, in bytes.
+        /// </summary>
+        public const int SharedSecretSizeInBytes = 32; // FIPS 203, Table 3.
```
I worry that FIPS 203-3 is going to introduce some new parameter set that when utilized emits a 48 byte shared secret; so all of my prototyping always had KEM SharedSecret size and all of the PrivateSeed sizes as `internal const` rather than `public const`.

Why would they do that? They did for DSA.

Why wouldn't they do that? They did for DSA, and it went terribly.

Hard to say if they've really learned their lesson :smile:

vcsjones replied:
> o all of my prototyping always had KEM SharedSecret size and all of the PrivateSeed sizes as internal const rather than public const.

I think it is really useful to expose in one way or another. Would you be opposed to it being on the `MLKemAlgorithm` instance? Or do you think it should just be internal for now?

bartonjs replied:
Since I was just reviewing FIPS 203 Table 3 (parameter set sizes), I see that NIST wrote `32` for each row, so they've clearly thought ahead to the time that it's not.

Seems like we should put it on the algorithm identifier class, then.

vcsjones replied:
That's true for the shared secret, but not the seed. Move both anyway?

bartonjs replied:
With just my API Review hat on, I'll ask

"Const is for all time.  Since you're building against an external spec you can't guarantee for all time.  Let's say this changes in 5 years... how do you react to that?"

If the answer is "we mark this as Obsolete and move it to MLKemAlgorithm", then we should probably just do that now.  If there's some other viable answer whose ugliness can be mitigated down by the belief it'll never happen... then I'm all ears 🐰

vcsjones replied:
Okay. Moving the concern go away, and there is no "harm" in moving it. So move it.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLKem.cs:
```diff
@@ -0,0 +1,453 @@
+                    nameof(sharedSecret));
+            }
+
+            if (ciphertext.Overlaps(sharedSecret))
```
Requiring non-overlapped seems fine to me.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLKem.cs:
```diff
@@ -0,0 +1,453 @@
+        /// <exception cref="ObjectDisposedException">The object has already been disposed.</exception>
+        public void Encapsulate(Span<byte> ciphertext, Span<byte> sharedSecret)
+        {
+            if (ciphertext.Length != Algorithm.CiphertextSizeInBytes)
```
I debated about this one.  I like that shared secret must be perfect, since it's the same for all ML-KEMs (currently :smile:), but was wondering if this should be != or <.  If it only throws for less-than, then we'd just call the Core method with ciphertext.Slice(0, Algorithm.CTSIB)... which is... fine (so the Core method always sees a perfect fit).

And an array-returning helper/caller would of course get this right.

So, is/was your belief that it should be a perfect fit here a conscious choice that the caller better show signs of understanding, or just a reflex reaction?

vcsjones replied:
> So, is/was your belief that it should be a perfect fit here a conscious choice that the caller better show signs of understanding, or just a reflex reaction?

It was intentional. I agree shared secret makes sense because it's the same, and I think it is weird for one to be one way and another to be the other.

I also think if we want to future proof for shared secret size being changeable then perhaps we need `out int ciphertextWritten, out int sharedSecretWritten`.

bartonjs replied:
Fair point about it being weird that one is exact and the other is not (which would change the void return to int return).  So, yeah, exact seems valid here.

> perhaps we need `, out int ciphertextWritten, out int sharedSecretWritten`.

We can add that as an overload.  If it's the only method, though, I expect people will just `out _` things, which means we made things more complicated for us and less usable for them.

So overload-soup might be

```C#
void Encapsulate(Span<byte> ciphertext, Span<byte> sharedSecret);
void Encapsulate(Span<byte> ciphertext, Span<byte> sharedSecret, out int ciphertextBytesWritten, out int sharedSecretBytesWritten);
byte[] Encapsulate(Span<byte> sharedSecret);
byte[] Encapsulate(Span<byte> sharedSecret, out int sharedSecretBytesWritten);
byte[] Encapsulate(out byte[] sharedSecret);
?
```

Maybe just saying "callers need to slice correctly" is the easiest.  And that's easy enough once we put SharedSecretSizeInBytes somewhere.

vcsjones replied:
Okay. Let's call this one "good" and I will make a note that the `out int` ones should exist when I get to overloadextravaganza?

vcsjones committed:
Remove Dispose checks from Core. The public APIs do that already

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/MLKem.cs:
```diff
@@ -0,0 +1,453 @@
+        protected abstract void ExportPrivateSeedCore(Span<byte> destination);
+
+        /// <summary>
+        /// Creates an ML-KEM key from a private seed.
```
Since Import is the name of the method (vs CreateFrom), this should probably use Imports in the doc comment.

For ML-DSA I wrote it as

```C#
/// <summary>
        ///   Imports an ML-DSA private key from its private seed value.
        /// </summary>
        /// <param name="algorithm">
        ///   The specific ML-DSA algorithm for this key.
        /// </param>
        /// <param name="source">
        ///   The bytes the key seed.
        /// </param>
        /// <returns>
        ///   The imported key.
        /// </returns>
        /// <exception cref="[CryptographicException](https://source.dot.net/System.Private.CoreLib/A.html#0b7aa8a209a7f19f)">
        ///   <para>
        ///     <paramref name="algorithm"/> is not a valid ML-DSA algorithm identifier.
        ///   </para>
        ///   <para>-or-</para>
        ///   <para>
        ///     <paramref name="source"/> is not the correct size for the specified algorithm.
        ///   </para>
        ///   <para>-or-</para>
        ///   <para>
        ///     An error occurred while importing the key.
        ///   </para>
        /// </exception>
```

The exception on algorithm being unknown is gone with the change from strongly typed string to the classes.

ArgumentException for length mismatch seems legit, though we should make sure that the signing algorithms treat it the same way (I'll take a note).  But you're also missing the good old "CryptographicException if something else goes wrong" ("An error occurred while importing the key")

vcsjones committed:
Code review feedback

vcsjones committed:
Move seed size to MLKemAlgorithm

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Encapsulation.cs:
```diff
@@ -0,0 +1,253 @@
+    public static partial class MLKemTests
+    {
+        [ConditionalFact(typeof(MLKem), nameof(MLKem.IsSupported))]
+        public static void DecapsulateVectors()
```
Yeah. I can figure out how to parameterize that when the work arrives.

vcsjones committed:
Fixup docs from code review

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Encapsulation.cs:
```diff
@@ -0,0 +1,253 @@
+        {
+            get
+            {
+                yield return new
```
I feel like in other algorithms when we are excerpting the CAVP/ACVP tests we cite them per vector-set.  More or less just indicating a difference between "this came from a standard test set" and "I did this using the openssl commandline" (or whatever).

vcsjones replied:
Ah, I did that for the key gen, but not for encapsulation.

vcsjones committed:
Add comment about test vector source

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Keys.cs:
```diff
@@ -0,0 +1,698 @@
+            }
+        }
+
+        [ConditionalTheory(nameof(IsNotSupported))]
```
Personally, I'm annoyed at how many "skipping test [blah] because [blah]" the test run prints.

I vaguely wonder if we should move the "not supported" tests to a different class and give it the class level attribute, and similarly make these have a class level attribute saying supported is required.

It's a minor thing, though, so can always do it later.

vcsjones replied:
Fair. I'll do it when we get to the "okay now make it work for `MLKemOpenSsl` and `MLKemCng`".

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Keys.cs:
```diff
@@ -0,0 +1,698 @@
+                    MLKemAlgorithm.MLKem768,
+                    "DB0686C838AEF0DC36363AAB0579ED25CA9BC6014CCAF5EE0E6B0783E4C9877EEC7D3B2991DBD5BC973E0F4362DF9B94F2760AA14AA2C8173017A30D6A871B45",
+                    "90C688F598CE9C15B4DB5C652C50088E24A6F9052167BA06A6937C6B58557578BBFA36C82F9B49701A9638C2CC716C91CA43643AF33EEEDC1B20B4B385C1924ABCC9D8419BC4723379628EB731CF99E50E598074F8332B3EF0A48A95477777972BA114D36A0B91CB0C9C2B07AB6B049F815B5B7731BC962C58D7CBEDB1205DEC3A5BD74CDA1A7338B674251726EDF3B397B4A1D0853D1D87C353582BDFC04B1057ACDCEA2136963F44157F6EDBA7E396622AF58F78C9592F626A3FE0BC ...
+                    "FAF5BDE1C50F9A7CB86D10B227D62621D0635342C3EB710037E480686B037DE8B8D9322D76E8331C3511BA29794A6575CDFAC5FA1191D6182FBBCB2FEFB3694A20A9CC247AA784B705780D770884F5327D10EA52697B3ED58C8A9BAC6C0F671EBC86BB2BBC8B06D3A4E0D21DCDE90E14690722A03B47E22665427754A8AEBAE76FA4B34B5273C84800C04ABC16D7C1C5631623BB3ABACCD7873817344E31240E3330C23783CB20A948B49ED71B395C05AD08996868A736E8C6B04215AA ...
```
OOh, a value starting with 0x00 :smile:

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MLKemImplementation.OpenSsl.cs:
```diff
@@ -0,0 +1,122 @@
+        private SafeEvpPKeyHandle _key;
+
+        // OpenSSL is expected to give "all or none" support.
+        internal static new bool IsSupported => Interop.Crypto.EvpKemAlgs.MlKem512 is not null;
```
My ML-DSA stub made a MLKemImplementation.cs that just defined everything as partial, so that errors are more clearly "this method is missing" rather than "overload resolution couldn't figure out what to call" (and it gives a nice thing to copy/paste from when setting up a new platform).  As part of that I ended up renaming IsSupported in the impl, either because `new partial` didn't work, or because it felt icky.

It doesn't look like you used the partial decl file for KEM.  I'm open to criticism that it doesn't add much and we shouldn't have them, but since it currently means ML-KEM is structured differently than ML-DSA it seems worth discussing.

vcsjones replied:
I have some mixed feelings on this. On one hand, yes the partial is a nice forcing function to make sure the statics are present (though the compiler will do that anyway, just with a slightly confusing error).

On the other hand, I value the lack of a partial. We use a lot of partials in the runtime and it can be hard to keep straight. I like "it's all one class in a file" and nice and tidy.

When partials bring value - like XML docs on public type - then I would absolutely agree that a partial is the right thing.

I don't mind adding it (too much) but its more lines of code and a file that doesn't _need_ to exist.

bartonjs commented on src/native/libs/System.Security.Cryptography.Native/osslcompat_30.h:
```diff
@@ -78,11 +81,25 @@ void EVP_MAC_free(EVP_MAC *mac);
+int EVP_PKEY_decapsulate(EVP_PKEY_CTX *ctx,
+    unsigned char *unwrapped, size_t *unwrappedlen,
+    const unsigned char *wrapped, size_t wrappedlen);
+    int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx, const OSSL_PARAM params[]);
+int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx, const OSSL_PARAM params[]);
+    int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
```
Indentation seems to have gotten weird here

bartonjs commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,302 @@
+{
+    return GetKeyOctetStringParam(pKey, OSSL_PKEY_PARAM_PRIV_KEY, destination, destinationLength);
+}
+
+
```
```suggestion

```

Nit, too many blank lines :smirk:

bartonjs commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_kem.c:
```diff
@@ -0,0 +1,302 @@
+                                       const uint8_t* ciphertext,
+                                       int32_t ciphertextLength,
+                                       uint8_t* sharedSecret,
+                                       int32_t sharedSecretLength)
```
This is the formatting that clang-format writes/wrote, right?

bartonjs commented on src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c:
```diff
@@ -775,3 +775,65 @@ EVP_PKEY_CTX* EvpPKeyCtxCreateFromPKey(EVP_PKEY* pkey, void* extraHandle)
+            goto done;
+        }
+
+        const char* paramName = privateKey == 0 ? OSSL_PKEY_PARAM_PUB_KEY : OSSL_PKEY_PARAM_PRIV_KEY;
+        int selection = privateKey == 0 ? EVP_PKEY_PUBLIC_KEY : EVP_PKEY_KEYPAIR;
+        size_t keyLengthT = Int32ToSizeT(keyLength);
```
Doesn't, at all, feel like Crypt32's dwMagic.  Nope.

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Keys.cs:
```diff
@@ -0,0 +1,698 @@
+
+            Assert.Throws<CryptographicException>(() => kem.ExportPrivateSeed(
+                new byte[MLKem.PrivateSeedSizeInBytes]));
+        }
```
I will add this to my scenario of "test cases when we have derived types" where "derived type = stub"

vcsjones committed:
Add key gen tests that have zero prefixes

vcsjones committed:
Add zero vectors for encap and decap

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Keys.cs:
```diff
@@ -0,0 +1,698 @@
+
+        public record class MLKemGenerateTestVector(MLKemAlgorithm Algorithm, string Seed, string EncapsulationKey, string DecapsulationKey);
+
+        // Generated from https://github.com/usnistgov/ACVP-Server/blob/85f8742965b2691862079172982683757d8d91db/gen-val/json-files/ML-KEM-keyGen-FIPS203/internalProjection.json
```
> Is this all of them, or is there a filtering criteria?

All of them. NIST's number of cases is somewhat restrained, so "why not".

For "weird", sure.

vcsjones committed:
Run clang-format

vcsjones committed:
Merge remote-tracking branch 'ms/main' into ml-kem

bartonjs commented on src/native/libs/System.Security.Cryptography.Native/osslcompat_30.h:
```diff
@@ -78,11 +81,25 @@ void EVP_MAC_free(EVP_MAC *mac);
+    const unsigned char *wrapped, size_t wrappedlen);
+int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx, const OSSL_PARAM params[]);
+int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx, const OSSL_PARAM params[]);
+    int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
+                             unsigned char *wrappedkey, size_t *wrappedkeylen,
+                             unsigned char *genkey, size_t *genkeylen);
```
EVP_PKEY_encapsulate still seems to be indented too far

```suggestion
int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
                         unsigned char *wrappedkey, size_t *wrappedkeylen,
                         unsigned char *genkey, size_t *genkeylen);
```

EVP_PKEY_fromdata and EVP_PKEY_encapsulate both use "align to (" for their newline, and EVP_PKEY_decapsulate is "indent one".

Feels like they should all be the same style

bartonjs reviewed (approved):
So long as that one header uses a style consistently, LGTM

vcsjones committed:
Fix formatting in header

vcsjones committed:
Merge remote-tracking branch 'ms/main' into ml-kem

vcsjones committed:
Actually fix the thing I said I fixed
